GitCommands/Submodules/SubmoduleStatusProvider.cs

```diff
@@ -21,7 +21,85 @@ public interface ISubmoduleStatusProvider : IDisposable
     }
 
-    public sealed class SubmoduleStatusProvider : ISubmoduleStatusProvider
+    public class SubmoduleStatusEventArgs : EventArgs
+    {
+        [NotNull]
+        public SubmoduleInfoResult Info { get; }
+
+        [NotNull]
+        public CancellationToken Token { get; }
+
+        public SubmoduleStatusEventArgs(SubmoduleInfoResult info, CancellationToken token)
+        {
+            Info = info;
+            Token = token;
+        }
+    }
+
+    public delegate void SubmoduleStatusEventHandler(object sender, SubmoduleStatusEventArgs e);
+
+    // Singleton provider
+    public sealed class SubmoduleStatusProvider
+    {
+        private ISubmoduleStatusProvider _impl = new SubmoduleStatusProviderImpl();
+
+        private SubmoduleStatusEventArgs _lastStatusEventArgs = null;
+
+        public static SubmoduleStatusProvider Default { get; } = new SubmoduleStatusProvider();
+
+        // Invoked when status update is requested (use to clear/lock UI)
+        public event EventHandler StatusUpdateBegin;
+
+        // Invoked when status update is complete
+        public event SubmoduleStatusEventHandler StatusUpdated;
+
+        public void Init()
+        {
+            _impl.Init();
+        }
+
+        public bool HasChangedToNone([CanBeNull] IReadOnlyList<GitItemStatus> allChangedFiles)
+        {
+            return _impl.HasChangedToNone(allChangedFiles);
+        }
+
+        public bool HasStatusChanges([CanBeNull] IReadOnlyList<GitItemStatus> allChangedFiles)
+        {
+            return _impl.HasStatusChanges(allChangedFiles);
+        }
+
+        public void UpdateSubmodulesStatus(bool updateStatus, string workingDirectory, string noBranchText)
+        {
+            _impl.UpdateSubmodulesStatus(updateStatus, workingDirectory, noBranchText, OnUpdateBegin, OnUpdateEndAsync);
+        }
+
+        public void ResendCachedStatus()
+        {
+            if (_lastStatusEventArgs != null)
+            {
+                OnUpdateBegin();
+                ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+                {
+                    await OnUpdateEndAsync(_lastStatusEventArgs.Info, _lastStatusEventArgs.Token);
+                }).FileAndForget();
+            }
+        }
+
+        private void OnUpdateBegin()
+        {
+            StatusUpdateBegin?.Invoke(this, EventArgs.Empty);
+        }
+
+        private async Task OnUpdateEndAsync(SubmoduleInfoResult info, CancellationToken token)
+        {
+            await TaskScheduler.Default;
+            token.ThrowIfCancellationRequested();
+            _lastStatusEventArgs = new SubmoduleStatusEventArgs(info, token);
+            StatusUpdated?.Invoke(this, _lastStatusEventArgs);
+        }
+    }
+
+    public sealed class SubmoduleStatusProviderImpl : ISubmoduleStatusProvider
```

Comment:
Yes, so there are two reasons I separated the two:

1. The "Impl" version does the actual work of retrieving the submodule status, while the singleton is a pubsub service responsible for calling back multiple subscribers, as well as caching the last status. I could mix the two together, but I think it might make it a bit harder to understand. I can easily make the Impl private, and/or give it a better name.

2. For practical reasons, having the singleton separate from the Impl allowed for me to easily merge changes that went in over the past few months. Of course, this is not a reason to leave it this way.

Reply:
@RussKie am going to try combining the two into one, as you suggested. Will also update the SubmoduleProviderTest that I wrote recently so that it builds and passes again.